gae_dashboard/waf_alert.py

```diff
@@ -0,0 +1,152 @@
+  100 * SUM(unique_messages)/SUM(total_messages) AS message_percentage,
+  APPROX_TOP_COUNT(b_request_user_agent,1)[OFFSET(0)].value AS request_user_agent,
+  100 * SUM(unique_ip)/SUM(total_ip) AS ip_percentage,
+  100 * SUM(unique_rua)/SUM(total_rua) AS rua_percentage
+FROM
+  BLOCKED_REQ
+  INNER JOIN TOTAL_REQ USING (t)
+WHERE
+  TIMESTAMP(t) BETWEEN TIMESTAMP('{start_timestamp}')
+  AND TIMESTAMP('{end_timestamp}')
+"""
+
+def waf_detect(end, ymd, waf_log_name, log_name):
+    start = end - datetime.timedelta(seconds=WAF_PERIOD)
+    query = QUERY_TEMPLATE.format(
+        fastly_log_tables=log_name,fastly_waf_log_tables=waf_log_name,
+        start_timestamp=start.strftime(TS_FORMAT),
+        end_timestamp=end.strftime(TS_FORMAT),
+        )
+    results = bq_util.query_bigquery(query, project=BQ_PROJECT)
+
+    # When an empty query is returned, these are of types ('None').
+    percentage = results[0]['percentage']
+    ip = results[0]['ip']
+    ip_percentage = results[0]['ip_percentage']
+    request_user_agent = results[0]['request_user_agent']
+    message = results[0]['message']
+    message_percentage = results[0]['message_percentage']
+    rua_percentage = results[0]['rua_percentage']
+
+    if ip_percentage == '(None)':
+        return
+
+    start_time = start.strftime("%H:%M")
+    end_time = end.strftime("%H:%M")
+
+    msg = """
+    :exclamation: *Possible WAF alert* :exclamation:
+    *Time Range:* {start_time} - {end_time}
+    *Blocked requests:* {percentage:.1f}% ({blocked} / {total})
+    *IP:* {ip} ({ip_percentage:.1f}% of blocks)
+    *Request User Agent:* {request_user_agent} ({rua_percentage:.1f}% of blocks)
+    *WAF Block Reason:* {message} ({message_percentage:.1f}% of blocks)
+    """.format(percentage=percentage, blocked=blocked, total=total, ip=ip, request_user_agent=request_user_agent, message=message, ip_percentage=ip_percentage, message_percentage=message_percentage, rua_percentage=rua_percentage, start_time=start_time, end_time=end_time)
+
+    if percentage > SPIKE_PERCENTAGE:
+        msg = """
+        :exclamation: *Possible WAF alert* :exclamation:
+        *Time Range:* {start_time} - {end_time}
+        *Blocked requests:* {percentage:.2f}%
+        *IP:* {ip} ({ip_percentage:.1f}% of blocks)
+        *Request User Agent:* {request_user_agent} ({rua_percentage:.1f}% of blocks)
+        *WAF Block Reason:* {message} ({message_percentage:.1f}% of blocks)
+        """.format(percentage=percentage, ip=ip, request_user_agent=request_user_agent, message=message, ip_percentage=ip_percentage, message_percentage=message_percentage, rua_percentage=rua_percentage, start_time=start_time, end_time=end_time)
+
+        alertlib.Alert(msg).send_to_slack(ALERT_CHANNEL)
+    else:
+        return
+
+def main():
+    # For demoing alert against specific attack time
+    # python waf_alert.py --endtime '2021-07-23 10:03:00'
+    # NOTE: Dates will only work for after 2021-07-20
```

Comment:
I think we should add this guard explicitly into the code to prevent accidentally running before this time.

```suggestion
    # NOTE: Dates will only work for after 2021-07-20
    assert now > datetime.date(2021, 7, 20), "Blocked column was introduced on 2021-07-20"
```